### ISSUE TITLE
Tasks don't seem to work with Terraform 2.1.0

### ISSUE THREAD

**Issue by jeanpaulsmit** [charleszipp/azure-pipelines-tasks-terraform#122]
I've been upgrading my Terraform definitions from 1.44 to 2.1 but executing them via the DevOps task results in an error about not being able to create the resource group.
This happens also with a brand new deployment.

_2020-03-23T11:55:08.1496745Z ##[section]Starting: Terraform apply
2020-03-23T11:55:08.1611757Z ==============================================================================
2020-03-23T11:55:08.1612183Z Task         : Terraform CLI
2020-03-23T11:55:08.1612493Z Description  : Execute terraform cli commands
2020-03-23T11:55:08.1612796Z Version      : 0.5.2
2020-03-23T11:55:08.1613079Z Author       : Charles Zipp
2020-03-23T11:55:08.1613348Z Help         : 
2020-03-23T11:55:08.1613692Z ==============================================================================
2020-03-23T11:55:08.7268235Z [command]C:\hostedtoolcache\windows\terraform\0.12.24\x64\terraform.exe version
2020-03-23T11:55:08.8218151Z Terraform v0.12.24
2020-03-23T11:55:09.2882701Z + provider.azurerm v2.1.0
2020-03-23T11:55:09.3031834Z 
2020-03-23T11:55:09.3172551Z [command]C:\hostedtoolcache\windows\terraform\0.12.24\x64\terraform.exe apply -auto-approve -var-file=api.dev.tfvars -var location=westeurope -var region=we
2020-03-23T11:55:15.1882553Z [0m[1mdata.azurerm_client_config.current: Refreshing state...[0m
2020-03-23T11:55:21.8110291Z [0m[1mazurerm_resource_group.rg: Creating...[0m[0m
2020-03-23T11:55:21.8771075Z [31m
2020-03-23T11:55:21.8776019Z [1m[31mError: [0m[0m[1mA resource with the ID "/subscriptions/7abcfb8e-0ghi-463b-b913-aadf518def26/resourceGroups/api-dev-we" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_group" for more information.[0m
2020-03-23T11:55:21.8778851Z 
2020-03-23T11:55:21.8781526Z [0m  on api.tf line 30, in resource "azurerm_resource_group" "rg":
2020-03-23T11:55:21.8782286Z   30: resource "azurerm_resource_group" "rg" [4m{[0m
2020-03-23T11:55:21.8782698Z [0m
2020-03-23T11:55:21.8782998Z [0m[0m
2020-03-23T11:55:21.9200429Z 
2020-03-23T11:55:21.9316911Z ##[error]Terraform command 'apply' failed with exit code '1'.
2020-03-23T11:55:22.5390425Z ##[section]Finishing: Terraform apply_

**Comment by charleszipp** [comment]
@jeanpaulsmit The error is indicating the resource already exists. Does it? If it does are you expecting it be recognized as already existing and not need to be recreated? If thats what you are expecting to happen and its not, im wondering if your backend/state configuration got changed during the migration such that you are now using a new/empty backend/state file?

Basically, this error indicates to me that your state/backend is out of sync with your deployment. The tasks responsibility is simply to string the appropriate command together and let terraform do the rest. Unless there is something wrong with the apply command it issued to terraform, this is unlikely to be an issue with the task itself.

**Comment by jeanpaulsmit** [comment]
Hi @charleszipp, thanks for the prompt response. The error is clear, but the interesting thing is if I change the azurerm version from 2.1 back to 1.44 then the entire script works flawlessly.

Actually this is the second time I ran into this, so I decided to drop a note. If I'm the only one having this issue, then it must be on my side though.

**Comment by jeanpaulsmit** [comment]
Running locally from a command prompt on the same subscription, but with storing the backend locally works fine with 2.1. 
That's why it took me some time to find out the cause as it only showed up when I ran it from the release pipeline and the error message doesn't really point in a useful direction.

**Comment by jeanpaulsmit** [comment]
It seems we need to run Terraform import for this to work.
@charleszipp Any idea when that will be supported?

**Comment by charleszipp** [comment]
> It seems we need to run Terraform import for this to work.
> @charleszipp Any idea when that will be supported?

Hey @jeanpaulsmit would you mind explaining why terraform import is necessary to migrate to the 2.0 version of the azurerm provider? I am looking through the upgrade guide on terraform's documentation for the provider and am unable to see any mention of this as a path to upgrade.

I have not encountered this situation myself so understanding why its needed would help me on two fronts. 1) make sure we get a solution implemented that actually addresses the underlying issue 2) make sure its implemented to solve for the most common use cases.

For example, are there breaking changes in state storage that require it be created by re-importing all existing resources?

I will try to reach out to some people on my end to see if others are encountering this within our team. Maybe they have a workaround that can be achieved with the current version.

**Comment by charleszipp** [comment]
@favoretti would be curious to get your thoughts on this? Perhaps you guys started with 2.0 version?

**Comment by jeanpaulsmit** [comment]
Hi @charleszipp we tried to upgrade from 1.44 to 2.x and came across the issue mentioned here: https://www.terraform.io/docs/providers/azurerm/guides/2.0-upgrade-guide.html#changes-when-importing-existing-resources
The "terraform init" task creates the resouce group and storage account to hold the tfstate file, but in the next step "apply" the behaviour is different in 2.0.
With the same tf file, the 2.0 version of azurerm errors with the resource group already exists and needs to be imported into the tfstate file.
This is part of the CLI but not of your task yet.
Not being able to import the resource group into the tfstate file, blocks us from moving to 2.x

**Comment by charleszipp** [comment]
Thanks for the explanation @jeanpaulsmit. I believe i have a better understanding as to why this is needed now.

In your case it would appear that the resource group used for the backend storage account is also used in your terraform templates. Which means you are trying to use a resource group that wasnt originally created by the templates (instead its created by az cli by this task). Prior to 2.0, apply would immediately fold that resource group into your state file if it already existed (basically upserting it). Now with 2.0, this same operation fails. The workaround is to import the resource so the terraform believe's it created the resource and therefore can update it.

Is that accurate?

**Comment by jeanpaulsmit** [comment]
Yes, totally accurate.
I'm not sure whether the import is a work arround though, or just the 2.x way to approach this.

**Comment by charleszipp** [comment]
Thanks @jeanpaulsmit this helps me make sure we fix it to solve the immediate problem. Seems like at the very least we need to support import to target specific resources that may have fallen under this situation before 2.0. 

Since #67 is very old and there have been some significant changes in the code base since, I will likely abandon this PR and start a fresh one from the latest code base. I can merge in some components from the #67 that should still work to speed up the process.

I can't commit to a timeline but, I will try to report back here as I get a better handle on when you can expect this.

**Comment by charleszipp** [comment]
@jeanpaulsmit will address with #129

**Comment by favoretti** [comment]
@charleszipp Yeah, we moved to 2.x a long time ago, basically as soon as it became available and back then we didn't use AzDO yet. Generally speaking, if @jeanpaulsmit can run terraform from local workstation with remote backend, the import can be done in a half-automated way: do all the imports/moves/any other necessary state modifications from a workstation, and then let extension handle all the rest.

**Comment by jeanpaulsmit** [comment]
@favoretti yeah, from a local workstation 2.x works fine, but we run the entire thing from Azure DevOps release pipelines (up to 1.44 this works rather well), so introducing a local workstation is not going to work.
In that case we would prefer a Powershell step in between using the Terraform CLI to do the job.

**Comment by favoretti** [comment]
The import is a one-off during migration, right? Shouldn't be that big a deal, although I don't know how large your TF codebase is of course. Although Charles should add import quite soon. :)

**Comment by jeanpaulsmit** [comment]
In the development phase we regularly remove the entire resource group and run the release pipeline again for a refresh. This however even should be possible for DR on production, so I rather have no manual steps unless it's absolutely necessary.

**Comment by charleszipp** [comment]
@jeanpaulsmit understand the situation. I typically do the same for the projects I work on to validate feature branches & PR's. 

I dont presume to know the situation or why certain decisions were made but, having the resource group that contains the backend storage account also be managed by your terraform templates is not a practice that i have seen typically done. Typically that backend and containing resource groups are completely isolated from the tf infrax deployment. I believe the reason others do this (and why i typically follow this pattern) is to ensure the backing state is protected from being manipulated, or worse, deleted by a defect in the templates. Something you may consider if you haven't already.

Regardless, I will be adding support for import as I have seen a few other cases internally where it is needed so I believe it will be a helpful addition to the task.

**Comment by charleszipp** [comment]
Confirmed theory stated above and what @jeanpaulsmit reported. Was able to reproduce the error @jeanpaulsmit stated above. Steps to repro...

1. Create a resource group `rg-tffoo-dev-eastus` via az cli or portal
2. Run terraform init on the sample templates from this repo @ `./TerraformTemplates`
3. Run apply on the same templates using the `./TerraformTemplates/default.vars` var file

Terraform will report the following error

```text
Error: A resource with the ID "/subscriptions/{my-subscription-id}/resourceGroups/rg-tfoo-dev-eastus" already exists - to be managed via Terraform this resource needs to be imported into the State. Please see the resource documentation for "azurerm_resource_group" for more 
information.

  on sample.tf line 1, in resource "azurerm_resource_group" "rg":
   1: resource "azurerm_resource_group" "rg" {
```

Will use the import command being implemented in #129 to confirm that will resolve the error in this case.

**Comment by jeanpaulsmit** [comment]
Thanks @charleszipp 

I started with a separate and central place to store the tfstate files, but switched to having it in the resource group it applies to, because it's very easy (and riskless) to entirely remove the deployment by just deleting the entire resource group.
Otherwise you end up with a large number of storage accounts in a particular resource group (or even worse: a single resource group with a bunch of containers). In order to remove a deployment we would need to pick the right tfstate file, which introduces a risk of selecting the wrong one or worse, someone deletes the entire storage account (or resource group).
It felt like to big of a risk to have the entire environment depending on what's in a single resource group or storage account.
Couldn't find much about the best practices when I started though.
